File: SportApplication/src/main/java/de/szut/sportapplication/model/Exercise.java
package de.szut.sportapplication.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Exercise {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "exercise_id")
    private Integer exerciseID;

    private String name;
    private String description;
    //TODO:gif und has weight noch hinzufügen/sowie spaltenname -- funktioniert in postman aber bis jetzt auch so
    // create und update?
    @Column(name = "has_weights")
    private boolean hasWeights;
    private String gifLink;
    private Integer xpTotal = 0;
    private Integer xpChest = 0;
    private Integer xpBack = 0;
    private Integer xpShoulders = 0;
    private Integer xpLegs = 0;
    private Integer xpTriceps = 0;
    private Integer xpAbs = 0;
    private Integer xpGlutes = 0;
    private Integer xpBiceps = 0;
    private Integer xpFlexibility = 0;

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public boolean isHasWeight() {
        return hasWeights;
    }

    public void setHasWeight(boolean hasWeight) {
        this.hasWeights = hasWeight;
    }

    public String getGifLink() {
        return gifLink;
    }

    public void setGifLink(String gifLink) {
        this.gifLink = gifLink;
    }

   // @OneToMany(mappedBy = "exercise", cascade = CascadeType.ALL, orphanRemoval = true)
   // private List<WorkoutExercise> workoutExercises = new ArrayList<>();

    public Integer getExerciseID() {return exerciseID;}

    public void setExerciseID (Integer exerciseID) {this.exerciseID = exerciseID;}

    public String getName() {
        return name;
    }
    public void setName(String name) {this.name = name;}
    public Integer getXpTotal() {return xpTotal;}
    public void setXpTotal(Integer xpTotal) {this.xpTotal = xpTotal;}
    public Integer getXpChest() {return xpChest;}

    public void setXpChest(Integer xpChest) {
        this.xpChest = xpChest;
    }
    public Integer getXpBack() {return xpBack;}
    public void setXpBack(Integer xpBack) {this.xpBack = xpBack;}
    public Integer getXpShoulders() {return xpShoulders;}
    public void setXpShoulders(Integer xpShoulders) {this.xpShoulders = xpShoulders;}
    public Integer getXpLegs() {return xpLegs;}
    public void setXpLegs(Integer xpLegs) {this.xpLegs = xpLegs;}
    public Integer getXpTriceps() {return xpTriceps;}
    public void setXpTriceps(Integer xpTriceps) {this.xpTriceps = xpTriceps;}
    public Integer getXpAbs() {return xpAbs;}

    public void setXpAbs(Integer xpAbs) {
        this.xpAbs = xpAbs;
    }
    public Integer getXpGlutes() {return xpGlutes;}
    public void setXpGlutes(Integer xpGlutes) {this.xpGlutes = xpGlutes;}
    public Integer getXpBiceps() {return xpBiceps;}
    public void setXpBiceps(Integer xpBiceps) {this.xpBiceps = xpBiceps;}
    public Integer getXpFlexibility() {return xpFlexibility;}
    public void setXpFlexibility(Integer xpFlexibility) {this.xpFlexibility = xpFlexibility;}

}

File: SportApplikation-FrontEnd/src/models/exercise.ts
export class Exercise {
  exerciseID: number;
  name: string;
  description: string;
  hasWeights: boolean;
  gifLink: string;
  xpTotal: number;
  xpChest: number;
  xpBack: number;
  xpShoulders: number;
  xpLegs: number;
  xpTriceps: number;
  xpAbs: number;
  xpGlutes: number;
  xpBiceps: number;
  xpFlexibility: number;
  numSets: number;
  numReps: number;
  weightAmount: number;

  //TODO: Constructor anpassen, erstmal für Testzwecke nur halb
  constructor(ExerciseID: number,Name: string,Description: string,gifLink: string, hasWeight:boolean, repetitions: number, sets:number, weight:number);
  constructor(ExerciseID: number,Name: string,Description: string,gifLink: string, hasWeight:boolean);
  constructor();
  constructor(
    ExerciseID?: number,
    Name?: string,
    Description?: string,
    gifLink?: string,
    hasWeight?:boolean,
    repetitions?:number,
    sets?:number,
    weight?:number,
  ){
  this.exerciseID = ExerciseID ?? 0;
  this.name = Name ?? "";
  this.description = Description ?? "";
  this.gifLink = gifLink ?? "";
  this.numReps = repetitions ?? 0;
  this.numSets = sets ?? 0;
  this.hasWeights = hasWeight ?? false;
  this.weightAmount = weight ?? 0;
  this.xpTotal = 0;
  this.xpChest = 0;
  this.xpBack = 0;
  this.xpLegs = 0;
  this.xpTriceps = 0;
  this.xpGlutes = 0;
  this.xpBiceps = 0;
  this.xpFlexibility = 0;
  this.xpShoulders = 0;
  this.xpAbs = 0;
}

  static fromJson(json: any): Exercise {
    const e = new Exercise();
    e.exerciseID = json.exerciseID;
    e.name = json.name;
    e.description = json.description;
    e.hasWeights = json.hasWeights;
    e.gifLink = json.gifLink;
    e.xpTotal = json.xpTotal;
    e.xpChest = json.xpChest;
    e.xpBack = json.xpBack;
    e.xpShoulders = json.xpShoulders;
    e.xpLegs = json.xpLegs;
    e.xpTriceps = json.xpTriceps;
    e.xpAbs = json.xpAbs;
    e.xpGlutes = json.xpGlutes;
    e.xpBiceps = json.xpBiceps;
    e.xpFlexibility = json.xpFlexibility;
    e.numSets = json.numSets ?? 0;
    e.numReps = json.numReps ?? 0;
    e.weightAmount = json.weightAmount ?? 0;

    return e;
  }

  static fromWorkoutExerciseJson(json: any): Exercise {
    const e = new Exercise();

    const exJson = json.exercise;

    e.exerciseID = exJson.exerciseID;
    e.name = exJson.name;
    e.description = exJson.description;
    e.hasWeights = exJson.hasWeights;
    e.gifLink = exJson.gifLink;
    e.xpTotal = exJson.xpTotal;
    e.xpChest = exJson.xpChest;
    e.xpBack = exJson.xpBack;
    e.xpShoulders = exJson.xpShoulders;
    e.xpLegs = exJson.xpLegs;
    e.xpTriceps = exJson.xpTriceps;
    e.xpAbs = exJson.xpAbs;
    e.xpGlutes = exJson.xpGlutes;
    e.xpBiceps = exJson.xpBiceps;
    e.xpFlexibility = exJson.xpFlexibility;
    e.numSets = json.numSets ?? 0;
    e.numReps = json.numReps ?? 0;
    e.weightAmount = json.weightAmount ?? 0;

    return e;
  }
}

